pkg/providers/ironic.go

```diff
@@ -0,0 +1,200 @@
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/httpbasic"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+	"github.com/gophercloud/gophercloud/pagination"
+	"go.etcd.io/etcd/pkg/transport"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"time"
+)
+
+var (
+	sometimeout = 30
+)
+
+type ironicProviderConfig struct {
+	Endpoint string `json:"endpoint"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Image    string `json:"image"`
+}
+type ironicProvider struct {
+	config ironicProviderConfig
+	client *gophercloud.ServiceClient
+}
+
+func IronicProviderFactory(providerInfo string, secretData map[string][]byte) (Provider, error) {
+	config := ironicProviderConfig{
+		Username: "",
+		Password: "",
+		Endpoint: "https://172.22.0.3:6385",
+		Image:    "http://172.22.0.1/images/ofcir_image.qcow2",
+	}
+
+	if configJSON, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJSON, &config); err != nil {
+			return nil, err
+		}
+	}
+
+	client, _ := httpbasic.NewBareMetalHTTPBasic(httpbasic.EndpointOpts{
```

Comment:
Could be maybe more convenient to make this a singleton, since a new instance it's created for every reconcile loop?